source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,300 @@
+  bool is_completed_{false};
+  // number of completed inner verifier for an any/all verifier.
+  std::size_t number_completed_children_{0};
+};
+
+class ContextImpl : public Verifier::Context {
+public:
+  ContextImpl(Http::HeaderMap& headers, Verifier::Callbacks* callback)
+      : headers_(headers), callback_(callback) {}
+
+  Http::HeaderMap& headers() const override { return headers_; }
+
+  Verifier::Callbacks* callback() const override { return callback_; }
+
+  void cancel() override {
+    for (const auto& it : auths_) {
+      it->onDestroy();
+    }
+    auths_.clear();
+  }
+
+  // Get Response data which can be used to check if a verifier node has responded or not.
+  CompletionState& getCompletionState(const Verifier* verifier) {
+    return completion_states_[verifier];
+  }
+
+  // Stores an authenticator object for this request.
+  void storeAuth(AuthenticatorPtr&& auth) { auths_.emplace_back(std::move(auth)); }
+
+private:
+  Http::HeaderMap& headers_;
+  Verifier::Callbacks* callback_;
+  std::unordered_map<const Verifier*, CompletionState> completion_states_;
+  std::vector<AuthenticatorPtr> auths_;
+};
+
+// base verifier for provider_name, provider_and_audiences, and allow_missing_or_failed.
+class BaseVerifierImpl : public Verifier {
+public:
+  BaseVerifierImpl(const BaseVerifierImpl* parent) : parent_(parent) {}
+
+  void completeWithStatus(Status status, ContextImpl& context) const {
+    if (parent_ != nullptr) {
+      return parent_->onComplete(status, context);
+    }
+    return context.callback()->onComplete(status);
+  }
+
+  // Check if next verifier should be notified of status, or if no next verifier exists signal
+  // callback in context.
+  virtual void onComplete(const Status& status, ContextImpl& context) const {
+    auto& completion_state = context.getCompletionState(this);
+    if (!completion_state.is_completed_) {
+      completion_state.is_completed_ = true;
+      completeWithStatus(status, context);
+    }
+  }
+
+protected:
+  // The parent group verifier.
+  const BaseVerifierImpl* const parent_;
+};
+
+// Provider specific verifier
+class ProviderVerifierImpl : public BaseVerifierImpl {
+public:
+  ProviderVerifierImpl(const std::string& provider_name, const AuthFactory& factory,
+                       const JwtProvider& provider, const BaseVerifierImpl* parent)
+      : BaseVerifierImpl(parent), auth_factory_(factory), extractor_(Extractor::create(provider)),
+        provider_name_(absl::optional<std::string>{provider_name}) {}
+
+  void verify(ContextSharedPtr context) const override {
+    auto& ctximpl = static_cast<ContextImpl&>(*context);
+    auto auth = auth_factory_.create(getSuppiler(), provider_name_);
+    extractor_->sanitizePayloadHeaders(ctximpl.headers());
+    auth->verify(
+        ctximpl.headers(), extractor_->extract(ctximpl.headers()),
+        [=](const Status& status) { onComplete(status, static_cast<ContextImpl&>(*context)); });
```

Comment:
just use ctximpl  so it doesn't need to cast again.

Reply:
you asked me to pass in a shared ptr.
https://github.com/envoyproxy/envoy/pull/4101#discussion_r215718280

Reply:
Sorry,  I missed that one.  Yes, we have to pass in shared_ptr so context is valid during the call.